app/Main.hs

```diff
@@ -0,0 +1,151 @@
+-- My lsp for some reason prefers this when formatting
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Main (main) where
+
+import Control.Applicative (liftA2, (<|>))
+import Control.Monad (guard, (<=<))
+import Data.Either.Extra (maybeToEither)
+import Data.Foldable (foldl')
+import Data.List (foldl1')
+import Data.Map.Strict (Map, empty, findWithDefault, insert, (!?))
+import Data.Text qualified as Text
+import Data.Text.IO qualified as TextIO
+import System.Console.ANSI (Color (..), ColorIntensity (..), ConsoleLayer (..), SGR (..), clearScreen, setCursorPosition, setSGR)
+import Text.Read (readMaybe)
+
+type Coordinates = (Int, Int)
+
+data Player = Cross | Circle deriving (Eq, Show)
+
+showPlayerT :: Player -> Text.Text
+showPlayerT = Text.pack . show
+
+newtype Board = Board (Map Coordinates Player)
+
+showBoard :: Board -> Text.Text
+showBoard (Board board) = foldl' replaceExclamation unfilledBoard $ flip (findWithDefault " ") textBoard <$> coordinates
```

Comment:
There's no need for textual manipulations after the textual board is generated.
Generate the right text from the start.